yarn-project/sequencer-client/src/simulator/acvm_native.ts

```diff
@@ -0,0 +1,113 @@
+ * @param bytecode - The circuit buytecode
+ * @param workingDirectory - A directory to use for temporary files by the ACVM
+ * @param pathToAcvm - The path to the ACVm binary
+ * @returns The completed partial witness outputted from the circuit
+ */
+export async function executeNativeCircuit(
+  inputWitness: WitnessMap,
+  bytecode: Buffer,
+  workingDirectory: string,
+  pathToAcvm: string,
+) {
+  const bytecodeFilename = 'bytecode';
+  const witnessFilename = 'input_witness.toml';
+
+  // convert the witness map to TOML format
+  let witnessMap = '';
+  inputWitness.forEach((value: string, key: number) => {
+    witnessMap = witnessMap.concat(`${key} = '${value}'\n`);
+  });
+
+  // In case the directory is still around from some time previously, remove it
+  await fs.rm(workingDirectory, { recursive: true, force: true });
+  // Create the new working directory
+  await fs.mkdir(workingDirectory, { recursive: true });
+  // Write the bytecode and input witness to the working directory
+  await fs.writeFile(`${workingDirectory}/${bytecodeFilename}`, bytecode);
+  await fs.writeFile(`${workingDirectory}/${witnessFilename}`, witnessMap);
+
+  // Execute the ACVM using the given args
+  const args = [
+    `execute`,
+    `--working-directory`,
+    `${workingDirectory}`,
+    `--bytecode`,
+    `${bytecodeFilename}`,
+    `--input-witness`,
+    `${witnessFilename}`,
+    `--print`,
+  ];
+  const processPromise = new Promise<string>((resolve, reject) => {
+    let outputWitness = Buffer.alloc(0);
+    let errorBuffer = Buffer.alloc(0);
+    const acvm = proc.spawn(pathToAcvm, args);
+    acvm.stdout.on('data', data => {
+      outputWitness = Buffer.concat([outputWitness, data]);
+    });
+    acvm.stderr.on('data', data => {
+      errorBuffer = Buffer.concat([errorBuffer, data]);
+    });
+    acvm.on('close', code => {
+      if (code === 0) {
+        resolve(outputWitness.toString('utf-8'));
+      } else {
+        reject(errorBuffer.toString('utf-8'));
+      }
+    });
+  });
+
+  try {
+    const output = await processPromise;
+    return parseIntoWitnessMap(output);
+  } finally {
+    // Clean up the working directory before we leave
+    await fs.rm(workingDirectory, { recursive: true, force: true });
+  }
+}
+
+export class NativeACVMSimulator implements SimulationProvider {
+  private count = 0;
+  constructor(private workingDirectory: string, private pathToAcvm: string) {}
+  async simulateCircuit(input: WitnessMap, compiledCircuit: NoirCompiledCircuit): Promise<WitnessMap> {
+    // Execute the circuit on those initial witness values
+
+    // Decode the bytecode from base64 since the acvm does not know about base64 encoding
+    const decodedBytecode = Buffer.from(compiledCircuit.bytecode, 'base64');
+
+    // Provide a unique working directory so we don't get clashes with parallel executions
+    const directory = `${this.workingDirectory}/${this.count}`;
+    ++this.count;
```

Comment:
I'd go with a uuid or even `Math.rand()`. If we accidentally create two instances of the NativeACVMSimulator (eg when running tests in parallel) using incremental counters will cause clashes.

Reply:
Yeah that was daft. We even have `randomBytes()` in `foundation`.